### PR TITLE
Personal/gankoji/authorization

### PR DESCRIPTION
**Pull request by gankoji**
This PR adds the RolesGuard class and the `Roles` decorator. RolesGuard performs JWT validation for requests to guarded endpoints, looks the user up in the DB and gets the roles assigned to that user, and validates that the user has the roles required for the request. The `Roles` decorator specifies metadata for each endpoint/query/mutation as to which roles are required to be able to successfully use said endpoint. Something like this:

```typescript
import {Roles} from "./roles.decorator";
import {RolesGuard} from "./roles.guard";

@Controller()
export class OauthController {
    constructor(private authService: OauthService) {}

    @Get("rolesAdminTest")
    @UseGuards(RolesGuard)
    @Roles("admin")
    getAdminRoles(): string | undefined {
        return "Wow you're an admin! Cool!";
    }
```
Role definitions are in the UserRolesType enum, and are currently placed directly into the User table. Refactor for Sov's "DB polymorphism" scema coming soon.

### CODE REVIEW

File: src/identity/auth/oauth/roles.guard.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            return req.query.token[0] as string;
+        }
+        return "";
+    }
+
+    async canActivate(context: ExecutionContext): Promise<boolean> {
+        const roles = this.reflector.get<string[]>("roles", context.getHandler());
+        /* eslint-disable */
+        if (!roles) {
+            return true;
```
Comment (andreasmwenzel) [inline src/identity/auth/oauth/roles.guard.ts:43]: is this the anonymous access case?

File: src/identity/auth/oauth/roles.guard.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        }
+        /* eslint-enable */
+        const request: Req = context.switchToHttp().getRequest();
+        const token = this.fromHeaderOrQueryString(request);
+        const jwtPayload: AuthPayload = this.jwtService.decode(token) as AuthPayload;
+        if (jwtPayload.sub && super.canActivate(context) as boolean) {
```
Comment (andreasmwenzel) [inline src/identity/auth/oauth/roles.guard.ts:49]: this spits out a big `TypeError: Cannot read properties of null (reading 'sub')` when an invalid jwtPayload is submitted

File: src/identity/auth/oauth/roles.guard.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        } else if (req.query.token) {
+            return req.query.token[0] as string;
+        }
+        return "";
+    }
+
+    async canActivate(context: ExecutionContext): Promise<boolean> {
+        const roles = this.reflector.get<string[]>("roles", context.getHandler());
+        /* eslint-disable */
+        if (!roles) {
```
Comment (andreasmwenzel) [inline src/identity/auth/oauth/roles.guard.ts:42]: can we make this work without upsetting eslint? I think it works as intended: roles doesn't exist on context handler -> allow, empty set of roles in handler -> will pass onto roles check which will deny it. But it's a really wonky typing interaction

Comment (gankoji) [reply src/identity/auth/oauth/roles.guard.ts:42]: I've tried. ESLint refuses to accept that roles could possibly be empty. Hence the disable bookends here.

**Review by 0c370t — APPROVED** [review]

**Pull request merged**